### DISCUSSION TITLE
Fiber Patch Panels

### DISCUSSION THREAD

**Discussion by deadlance** [netbox-community/netbox#17530] [Help Wanted!]
I am using Netbox to track connections and all sorts of things that are maybe outside the norm.

Facility A - I have a rack, call it Rack01. In Rack01, I have a 4U fiber patch panel that holds cassettes for fiber terminations. There are 288 strands of singlemode fiber terminated in that panel.

Facility B - I have a rack, call it Rack02. Rack02 has a fiber patch panel that is exactly the same as the panel in Rack01.

What's the best way of connecting these panels so that when I trace an interface plugged into this panel, it correctly shows up all the way through?

Currently, I am connecting each rear port individually. Rack01, Cassette A, port 1 connects to rear port, rack02, cassette A, port 1, and so on. If I connect them in bulk IE Rack01 A1-M24 to Rack02 A1-M24, it always looks like any interface traced just goes through the same cable, which is true, but there are other issues. For example, when I attempt to connect a network switch interface to Rack01 A1:2, it throws an assertion error.

So, the end result question is, what's the fastest and most appropriate way to connect Rack01 to Rack02, all 288 strands, but still be able to connect a switch interface to Rack01 A1:2 and it trace through to Rack02 A1:2. Currently I am connecting these 1 strand at a time (rear port to rear port) and 288 strands takes forever.

**Comment by candlerb** [comment]
> If I connect them in bulk IE Rack01 A1-M24 to Rack02 A1-M24, it always looks like any interface traced just goes through the same cable, which is true, but there are other issues.

I don't know quite what you mean by "connect in bulk", but if you mean you have a single cable with 24 terminations at each end, then that *won't* keep track of the correct port relationships when tracing; the terminations are all treated as equivalent. Indeed, it's not even required to have the same number of terminations at each end.

(The main use case for this functionality was to be able to have a duplex fibre cable, connected to two patch panel ports at one end, and a single interface at the other end, forming a single logical connection - not to trace individual "strands" through a multi-core cable)

To get the functionality you want, so you can trace all the way through, you have two main choices:

1. Create patch panels or cassettes with a single rear port associated to multiple front ports, and then connect the rear ports together with a single cable. This will trace fine: Netbox keeps track of the strand offset within the bulk cable while tracing, even though the cable is modelled as a single connection with a single termination at each end.

2. Associate each front port to a single rear port, and connect all the rear ports together 1:1 with *separate* cables. A script like [this one](https://github.com/netbox-community/customizations/blob/master/scripts/multi_connect.py) can help you create all the different cables. ~~Note that there is an outstanding [pull request](https://github.com/netbox-community/customizations/pull/107) for that script to make it work with Netbox 4.0+, which has been waiting merge since May.~~ [FIXED]

Which option you use depends on how you use your fibre infrastructure.

**Option 1: Pairs**
If it's 100% duplex pairs, then the simplest approach is to treat a pair of fibres as a single entity, where a single frontport represents a pair of optical connectors (e.g. a pair of LC or SC). Your 288 fibre strands would present as 144 front ports.

Then you create each 24-strand cassette as having 12 front ports and 1 rear port. You connect the rear port from one cassette to the rear port of a corresponding cassette in a different rack. You then model patch cables as being a single cable with one termination at each end (e.g. one interface to one front port).

This works very well, but only if you never need to use the two strands of a pair separately. It means you can ignore how the TX/RX pairs are being crossed over, which is commonly 3 times (once in the patch cable at each end, and once between the patch panels)

**Option 2: Single strands**
If you use single strands, e.g. for Bidi or GPON, then you'll want to model each strand and front port separately. Connecting an interface to a patch panel would be a cable with a single termination at one end (into the interface), and two terminations at the patch panel end (into ports 1 and 2, say).

You could have 24 front ports associated to 1 rear port on a cassette as above: *but* that requires the ports to be connected one-to-one. If you are using MPO cables with crossovers (1-2, 2-1, 3-4, 4-3 etc) then this won't work, since Netbox can't handle anything other than a 1:1 relationship between the front ports at each end. (To find out if this is the case, shine a VFL into port 1 on a patch panel, and see if it lights up port 1 or port 2 at the other end)

In this case, you're forced to model each front port connected to a single rear port, and have individual cables between the rear ports, crossed over appropriately.

Final point to note: if you're using modular cassettes, like the [Complete Connect MX](https://www.completeconnect.co.uk/mx-xpress-pre-terminated-modular-fibre/) series, then I'd be inclined to model the individual cassettes and the 1U or 2U frames which house them separately. You can do this by modelling the frame as a "parent device" with device bays, and each cassette as "child device". This makes it clear which cassette slots are occupied or not, and gives you some flexibility about routing each cassette to a different physical location. However, it does make it harder to see all the connections in one go in the UI, since you have to drill down to each cassette. If you want to treat the whole thing logically as a 288-port patch panel, then that's fine too.

**Reply by jsenecal** [reply]
I just merged your contribution @candlerb, sorry that it took so long.

**Reply by candlerb** [reply]
Appreciated, thank you!

**Comment by deadlance** [comment]
I am currently using your Option 2. 

I have a patch panel that is a mirror of another patch panel. I was wondering if there was a way to quickly connect those 2 panels using your option 2.

I have modeled a patch panel that accepts "modules" and have cassettes modeled as modules to plug into those modules in the patch panel.  Once I drop a patch panel in a location, I can easily plugin the cassette modules, but then I have to connect the rear port to the opposite end.

The situation might be more easily explained - I have 2 buildings, 20 miles apart. I am running a 288 SMF between the 2 buildings, and at each location, the 288 will be terminated into a 4U patch panel, using cassettes A through M. That part is easy and already done. The tedious part is connecting those in netbox. What I'd like to do is be able to build out a spreadsheet or write a script that just says Panel A, strands 1-288 connect to Panel B, strands 1-288 where the ports match... or something like that instead of modeling each strand individually. Currently connecting all 288 strands, 1 at a time, takes a few hours.

It all feels like overkill except I need to have the ability to click the trace cable button and see every connection, port, cable length, glass through site, etc., between locations.

**Reply by candlerb** [reply]
> What I'd like to do is be able to build out a spreadsheet or write a script that just says Panel A, strands 1-288 connect to Panel B, strands 1-288 where the ports match

The script I linked to will do that for you (within Netbox). Or you can generate a CSV file with the connections, and import it.

However, if all the strands are connected 1:1 and not crossed over, you could go with option 1 and reduce the number of cables by a factor of 24.
